src/main/java/com/google/devtools/build/lib/rules/repository/LocalRepositoryFunction.java

```diff
@@ -117,21 +117,7 @@ public boolean isLocal(Rule rule) {
   @Nullable
   protected static FileValue getWorkspaceFile(RootedPath directory, Environment env)
       throws RepositoryFunctionException, InterruptedException {
-    RootedPath workspaceRootedFile =
-        RootedPath.toRootedPath(
-            directory.getRoot(),
-            directory
-                .getRootRelativePath()
-                .getRelative(LabelConstants.WORKSPACE_DOT_BAZEL_FILE_NAME));
-
-    if (!workspaceRootedFile.asPath().exists()) {
-      workspaceRootedFile = RootedPath.toRootedPath(
-          directory.getRoot(),
-          directory
-              .getRootRelativePath()
-              .getRelative(LabelConstants.WORKSPACE_FILE_NAME));
-    }
-
+    RootedPath workspaceRootedFile = WorkspaceFileHelper.getWorkspaceRootedFile(directory);
```

Comment:
I just realized you must not use Path.exists() in SkyFunctions. You must request the FileValue for "WORKSPACE.bazel", check its type, and only if it's missing or not a regular file, then request FileValue for "WORKSPACE".

This way the SkyFunction will correctly track changes on the filesystem (e.g. if the file "WORKSPACE.bazel" is replaced by a directory of the same name -- actually, could you add a test for that?)

 think this logic is incorrect: it circumvents Skyframe. You have to ask Skyframe about the file existing, you mustn't use Path.exists().

For taht, request the FileValue for "WORKSPACE.bazel", and check is type: if it's not an existing file, then get the FileValue for "WORKSPACE".
This way Skyframe correctly tracks changes to these files and in the directory.

Reply:
Got it! Thanks for reminding me of that! This is extremely important!

Reply:
I had to move WorkspaceFileHelper.java to another package to avoid dependencies problem.